### PR TITLE
DL-06: Improve chatbot intent classification

### PR DESCRIPTION
**Pull request by Sreya256**
## Overview

Implemented the natural-language intent detection component for the DiscountMate chatbot as part of DL-06-T5.

## Contributions

- Created and labelled 173 natural-language chatbot queries.
- Covered all seven supported chatbot intents:
  - `price_comparison`
  - `product_search`
  - `product_details`
  - `retailer_search`
  - `category_search`
  - `product_substitution`
  - `no_action`
- Added variations in natural-language query structures and relevant shopping entities.
- Implemented a TF-IDF + Logistic Regression baseline classifier.
- Used a stratified 75/25 train-test split for model evaluation.
- Added classification report and confusion matrix analysis.
- Added an automated unseen-query test suite.
- Added documentation covering the dataset, approach, evaluation, limitations and future improvements.

## Evaluation

Current held-out evaluation:

- Test examples: 44
- Accuracy: 75.00%
- Macro F1: 0.75
- Weighted F1: 0.74

Separate unseen-query validation:

- Total tests: 14
- Passed: 14
- Failed: 0
- Validation accuracy: 100%

The 100% result refers specifically to the selected unseen validation queries and is not the overall model accuracy.

## Limitations

The classifier is currently a baseline intent-detection component. Some intents have overlapping language and some predictions have relatively low confidence.

Entity extraction, database querying and complete chatbot/API integration are outside the current T5 implementation and can be addressed in subsequent DL-06 tasks.

## Testing

Run the following from `Backend/ml-service`:

```bash
python intent_classifier.py
python test_intent_classifier.py

### CODE REVIEW

**Comment by callmesoffie1811** [comment]
Hi @Sreya256, I’ve reviewed PR #330. The intent-classification implementation looks well scoped as a baseline, with the seven intents, labelled dataset, TF-IDF + Logistic Regression pipeline, evaluation and test queries included.

Before merge, could you please confirm the model has been evaluated and is performing as expected for the intended showcase use case? If possible, it would be helpful to include the evaluation results or a few representative predictions.

Given the showcase timeline, I don’t think we need to expand the model or dataset further at this stage. If the current results are considered suitable for the baseline, I’m happy for it to proceed through the normal approval process.

**Comment by Sreya256** [comment]
> Hi @Sreya256, I’ve reviewed PR #330. The intent-classification implementation looks well scoped as a baseline, with the seven intents, labelled dataset, TF-IDF + Logistic Regression pipeline, evaluation and test queries included.
> 
> Before merge, could you please confirm the model has been evaluated and is performing as expected for the intended showcase use case? If possible, it would be helpful to include the evaluation results or a few representative predictions.
> 
> Given the showcase timeline, I don’t think we need to expand the model or dataset further at this stage. If the current results are considered suitable for the baseline, I’m happy for it to proceed through the normal approval process.

**Comment by Sreya256** [comment]
Hi @callmesoffie1811 , thanks for reviewing the PR.

I can confirm that the intent classifier has been evaluated for the intended showcase use case. The current held-out evaluation uses a stratified 75/25 train-test split with 44 test examples and achieved 75% accuracy, 0.75 macro F1, and 0.74 weighted F1.

I also tested a separate set of 14 selected unseen queries covering all seven intents, with 14/14 correct predictions (100% validation accuracy). This result refers specifically to the selected unseen-query validation set and is not the overall model accuracy.

Representative predictions include:

“Which supermarket has Coke Zero at the lowest price?” → price_comparison
“Find some snack products.” → category_search
“What's a cheaper alternative to Milo?” → product_substitution
“What's the weather like today?” → no_action

These results are considered suitable for the baseline showcase use case. I have also documented the evaluation results and limitations in the README. Thanks!

**Comment by callmesoffie1811** [comment]
It looks good to me, the PR is ready to merge.

**Review by anhvdq — APPROVED** [review]
LGTM
